File: src/client/spritedata.cpp
/*
 * Copyright 2012 Scott MacDonald
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "client/spritedata.h"
#include "common/platform.h"

#include <QPixmap>
#include <QImage>
#include <QSize>

/**
 * Standalone sprite (one that is not located in a spritesheet) constructor
 */
SpriteData::SpriteData( const QPixmap& pixmap )
    : mPixmap( pixmap ),
      mSize( pixmap.size() )
{
}

/**
 * SpriteData constructor for a sprite that is located within a
 * spritesheet image. X and Y denote the offset, and width/height
 * denote how big the sprite is.
 *
 * \param  pTexture  Pointer to the texture holding the sprite sheet
 * \param  xOffset   Left origin of the sprite in the sprite sheet
 * \param  yOffset   Top origin of the sprite in the sprite sheet
 * \param  width     Width of the sprite
 * \param  height    Height of the sprite
 */
SpriteData::SpriteData( const QImage& pixmap,
                        int xOffset,
                        int yOffset,
                        int width,
                        int height )
    : mPixmap(),
      mSize( width, height )
{
    QImage chunk = pixmap.copy( xOffset, yOffset, width, height );
    mPixmap      = QPixmap::fromImage( chunk );
}

/**
 * Copy constructor
 */
SpriteData::SpriteData( const SpriteData& s )
    : mPixmap( s.mPixmap ),
      mSize( s.mSize )
{
}

/**
 * Destructor
 */
SpriteData::~SpriteData()
{
}

/**
 * Assignment operator.
 */
SpriteData& SpriteData::operator = ( const SpriteData& rhs )
{
    mPixmap = rhs.mPixmap;
    mSize  = rhs.mSize;

    return *this;
}

/**
 * Equality operator
bool SpriteData::operator == ( const SpriteData& rhs ) const
{
    return ( mPixmap == rhs.mPixmap && mSize == rhs.mSize );
}

/**
 * Inequality operator

bool SpriteData::operator != ( const SpriteData& rhs ) const
{
    return ( mPixmap != rhs.mPixmap || mSize != rhs.mSize );
}*/

/**
 * Return a pointer to the sprite's underlying texture object
 */
const QPixmap& SpriteData::pixmap() const
{
    return mPixmap;
}


/**
 * The sprite's width
 */
int SpriteData::width() const
{
    return mSize.width();
}

/**
 * The sprite's height
 */
int SpriteData::height() const
{
    return mSize.height();
}

File: src/Scott.Forge.Engine/Graphics/SpriteData.cs
/*
 * Copyright 2012-2013 Scott MacDonald.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
using System;
using System.Collections.Generic;
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;

namespace Scott.Forge.Engine.Graphics
{
    /// <summary>
    ///  SpriteData holds data describing 2d graphical image known as a "Sprite". This class is
    ///  responsible for holding basic sprite data (Such as name, width, height, etc), a list of
    ///  animations that the sprite is capable of performing and the information required for
    ///  displaying each image frame.
    /// </summary>
    /// <remarks>
    ///  SpriteData holds one or more images, and these images are stored in one or more texture
    ///  atlases. SpriteData describes how to index the images stored in the texture atlas.
    ///  
    ///  Generally SpriteData should not be modified at run time. It should be modified when it is
    ///  stored on disk, and then read into memory.
    ///  
    ///  TODO: Make this an internal class. Make a struct Sprite class.
    /// </remarks>
    /// [ContentImporter( ".sprite", DisplayName = "Sprite Xml Importer", DefaultProcessor = "SpriteContentProcessor" )
    public class SpriteData         // TODO: Rename SpriteDefinition
    {
        /// <summary>
        /// Name of the sprite
        /// </summary>
        public string Name { get; set; }

        /// <summary>
        /// The sprite's texture atlas
        /// </summary>
        public Texture2D Texture { get; set; }

        /// <summary>
        /// List of animations for this sprite
        /// </summary>
        public Dictionary<string, AnimationData> Animations { get; set; }

        public AnimationData DefaultAnimation { get; private set; }

        /// <summary>
        /// The default animation name for this sprite
        /// </summary>
        public string DefaultAnimationName { get; set; }

        /// <summary>
        /// Direction for the default animation
        /// </summary>
        public DirectionName DefaultAnimationDirection { get; set; }

        /// <summary>
        /// Offset from local origin when rendering
        /// </summary>
        public Vector2 OriginOffset { get; set; }

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="name">Name of the sprite</param>
        /// <param name="texture">Texture atlas that contains this sprite</param>
        /// <param name="defaultAnimation">Default animation name</param>
        /// <param name="defaultDirection">Default direction to use when animating without a direction</param>
        internal SpriteData( string name,
                             Texture2D texture,
                             string defaultAnimation,
                             DirectionName defaultDirection,
                             List<AnimationData> animationList )
        {
            Name = name;
            Texture = texture;
            
            // Copy animations
            Animations = new Dictionary<string, AnimationData>( animationList.Count );

            foreach ( AnimationData animation in animationList )
            {
                Animations.Add( animation.Name, animation );
            }

            DefaultAnimationName = defaultAnimation;
            DefaultAnimationDirection = defaultDirection;
            DefaultAnimation = Animations[DefaultAnimationName];
            OriginOffset = Vector2.Zero;
        }
    }
}
